### PR TITLE
Rou 3142

### PR DESCRIPTION
**Pull request by gnbm**
This PR adds a new API called **MarkChangesAsSavedByKey**, to mark a group of Data Grid rows with given keys (from the KeyBinding field) as saved in the database.

### Test Steps

**Validate client action MarkChangesAsSavedByKey with success scenario:**
1. Open Page
2. Add a new row by clicking in "Add New Row"
3. For the first row, change the content of column Product Name and a dirty mark will appear
4. Fill in the Row Keys values of "-1,45"
4. Press "Mark Changes As Saved By Key" button

Expected: 
- A success message should appear "Mark Changes As Saved successfully";
- No dirty mark is shown

**Validate client action MarkChangesAsSavedByKey with success scenario - Force Clean Invalids:**
1. Open Page
2. Add a new row by clicking in "Add New Row"
3. Click in "Set Cell Invalid"
4. Enable "Force Clean Invalids"
4. Fill in the Row Keys values of "-1,56"
4. Press "Mark Changes As Saved By Key" button

Expected: 
- A success message should appear "Mark Changes As Saved successfully";
- No dirty mark nor validation marks are shown

**Validate client action MarkChangesAsSavedByKey with empty rows list:**
1. Open Page
2. Press "Mark Changes As Saved By Key - Empty" button

Expected: 
- An error message should appear "Rows list is empty";

**Validate client action MarkChangesAsSavedByKey with rows list containing empty key values:**
1. Open Page
2. Press "Mark Changes As Saved By Key - With Empty Key" button

Expected: 
- An error message should appear "Rows list has empty values.";

**Validate client action MarkChangesAsSavedByKey with invalid row:**
1. Open Page
2. Press "Mark Changes As Saved By Key - Invalid Row" button

Expected: 
- An error message should appear "Row not found.";

### Checklist
* [ X] tested locally
* [ X] documented the code
* [ X] clean all warnings and errors of eslint
* [ X] requires changes in OutSystems (if so, provide a module with changes)
* [ X] requires new sample page in OutSystems (if so, provide a module with changes)

### CODE REVIEW

File: code/src/WijmoProvider/Grid/FlexGrid.ts [hunk 1]
```diff
@@ -186,6 +212,67 @@ namespace WijmoProvider.Grid {
...
+         * @param {boolean} forceClearValidationMarks determines whether or not we should clean the validation marks.
+         * @memberof FlexGrid
+         */
+        public clearAllChangesByRowKeys(
+            rowKeys: Array<string>,
+            forceClearValidationMarks: boolean
+        ): void {
+            //if the row keys array is empty we will throw an error
+            if (rowKeys.length === 0) {
+                throw new Error(OSFramework.Enum.ErrorMessages.Row_EmptyList);
```
Comment (joselrio) [inline code/src/WijmoProvider/Grid/FlexGrid.ts:228]: Missing ErrorCode?

File: code/src/WijmoProvider/Grid/FlexGrid.ts [hunk 2]
```diff
@@ -186,6 +212,67 @@ namespace WijmoProvider.Grid {
...
+            }
+            //if the row keys array as empty keys we will throw an error
+            if (
+                rowKeys.indexOf('') > -1 ||
+                rowKeys.indexOf(null) > -1 ||
+                rowKeys.indexOf(undefined) > -1
+            ) {
+                throw new Error(
+                    OSFramework.Enum.ErrorMessages.Row_ListEmptyValues
```
Comment (joselrio) [inline code/src/WijmoProvider/Grid/FlexGrid.ts:237]: Missing ErrorCode?

File: code/src/WijmoProvider/Grid/FlexGrid.ts [hunk 3]
```diff
@@ -186,6 +212,67 @@ namespace WijmoProvider.Grid {
...
+                            (item) =>
+                                _.get(
+                                    item.dataItem,
+                                    this.config.keyBinding
+                                ).toString() === element
+                        );
+
+                        if (row === -1) {
+                            throw new Error(
+                                OSFramework.Enum.ErrorMessages.Row_NotFound
```
Comment (joselrio) [inline code/src/WijmoProvider/Grid/FlexGrid.ts:259]: Missing ErrorCode?

File: code/src/WijmoProvider/Grid/FlexGrid.ts (hunk 1 above)
Comment (gnbm) [reply code/src/WijmoProvider/Grid/FlexGrid.ts:228]: @joselrio The error codes are set at the API level.

File: code/src/WijmoProvider/Grid/FlexGrid.ts (hunk 2 above)
Comment (gnbm) [reply code/src/WijmoProvider/Grid/FlexGrid.ts:237]: @joselrio The error codes are set at the API level.

File: code/src/WijmoProvider/Grid/FlexGrid.ts (hunk 3 above)
Comment (gnbm) [reply code/src/WijmoProvider/Grid/FlexGrid.ts:259]: @joselrio The error codes are set at the API level.

**Review by joselrio — APPROVED** [review]

File: code/extension/DataGridUtils/Source/NET/obj/Release/.NETFramework,Version=v4.6.1.AssemblyAttributes.cs [hunk 4]
```diff
@@ -0,0 +1,4 @@
+// <autogenerated />
```
[the remark this answers is not in the document]
Comment (gnbm) [reply code/extension/DataGridUtils/Source/NET/obj/Release/.NETFramework,Version=v4.6.1.AssemblyAttributes.cs]: Done

**Pull request merged**